{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopython 3: SearchIO and Entrez\n",
    "\n",
    "In this workshop, we will take a quick look at SearchIO, Biopython's tool for unifying results from various sequence search tools, before exploring the Entrez API utility from NCBI.\n",
    "\n",
    "## SearchIO\n",
    "\n",
    "Biopython provides the [SearchIO package](https://biopython.org/docs/latest/api/Bio.SearchIO.html) for dealing with outputs from various sequence searching utilities, allowing them to be compared directly. Similar to SeqIO and AlignIO, it has capabilities for parsing, reading, writing, and indexing different search results, as well as converting between file types. \n",
    "\n",
    "### Reading and Parsing\n",
    "\n",
    "Parsing and reading also work the same way as previously, where parse returns an iterator, and read simply returns the first item in the file. Note here that each item is a query, not an individual sequence. \n",
    "\n",
    "Let's start with an example of reading and XML result file from a BLAST search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['_NON_STICKY_ATTRS', '_QueryResult__alt_hit_ids', '_QueryResult__marker', '__annotations__', '__bool__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_description', '_hit_key_function', '_id', '_items', '_transfer_attrs', 'absorb', 'append', 'blast_id', 'description', 'fragments', 'hit_filter', 'hit_keys', 'hit_map', 'hits', 'hsp_filter', 'hsp_map', 'hsps', 'id', 'index', 'items', 'iterhit_keys', 'iterhits', 'iteritems', 'param_evalue_threshold', 'param_filter', 'param_gap_extend', 'param_gap_open', 'param_score_match', 'param_score_mismatch', 'pop', 'program', 'reference', 'seq_len', 'sort', 'stat_db_len', 'stat_db_num', 'stat_eff_space', 'stat_entropy', 'stat_hsp_len', 'stat_kappa', 'stat_lambda', 'target', 'version']\n",
      "Search 42291 has 100 hits\n",
      "['_NON_STICKY_ATTRS', '__annotations__', '__bool__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_aln_span_get', '_get_coords', '_hit_end_get', '_hit_inter_ranges_get', '_hit_inter_spans_get', '_hit_range_get', '_hit_span_get', '_hit_start_get', '_inter_ranges_get', '_inter_spans_get', '_items', '_query_end_get', '_query_inter_ranges_get', '_query_inter_spans_get', '_query_range_get', '_query_span_get', '_query_start_get', '_str_hsp_header', '_transfer_attrs', '_validate_fragment', 'aln', 'aln_all', 'aln_annotation', 'aln_annotation_all', 'aln_span', 'bitscore', 'bitscore_raw', 'evalue', 'fragment', 'fragments', 'gap_num', 'hit', 'hit_all', 'hit_description', 'hit_end', 'hit_end_all', 'hit_features', 'hit_features_all', 'hit_frame', 'hit_frame_all', 'hit_id', 'hit_inter_ranges', 'hit_inter_spans', 'hit_range', 'hit_range_all', 'hit_span', 'hit_span_all', 'hit_start', 'hit_start_all', 'hit_strand', 'hit_strand_all', 'ident_num', 'is_fragmented', 'molecule_type', 'output_index', 'pos_num', 'query', 'query_all', 'query_description', 'query_end', 'query_end_all', 'query_features', 'query_features_all', 'query_frame', 'query_frame_all', 'query_id', 'query_inter_ranges', 'query_inter_spans', 'query_range', 'query_range_all', 'query_span', 'query_span_all', 'query_start', 'query_start_all', 'query_strand', 'query_strand_all']\n",
      "gi|262205317|ref|NR_030195.1| 4.91307e-23\n",
      "gi|301171311|ref|NR_035856.1| 1.71483e-22\n",
      "gi|270133242|ref|NR_032573.1| 2.54503e-20\n",
      "gi|301171322|ref|NR_035857.1| 8.88303e-20\n",
      "gi|301171322|ref|NR_035857.1| 3.31332e-06\n",
      "gi|301171267|ref|NR_035851.1| 8.88303e-20\n",
      "gi|262205330|ref|NR_030198.1| 8.88303e-20\n",
      "gi|262205330|ref|NR_030198.1| 3.31332e-06\n",
      "gi|262205302|ref|NR_030191.1| 8.88303e-20\n",
      "gi|301171259|ref|NR_035850.1| 3.10048e-19\n",
      "gi|262205451|ref|NR_030222.1| 3.10048e-19\n",
      "gi|301171447|ref|NR_035871.1| 3.77716e-18\n",
      "gi|301171447|ref|NR_035871.1| 0.000140886\n",
      "gi|301171276|ref|NR_035852.1| 3.77716e-18\n",
      "gi|262205290|ref|NR_030188.1| 3.77716e-18\n",
      "gi|301171354|ref|NR_035860.1| 1.31836e-17\n",
      "gi|262205281|ref|NR_030186.1| 1.31836e-17\n",
      "gi|262205298|ref|NR_030190.1| 4.60152e-17\n",
      "gi|262205298|ref|NR_030190.1| 0.000140886\n",
      "gi|301171394|ref|NR_035865.1| 1.60609e-16\n",
      "gi|262205429|ref|NR_030218.1| 1.60609e-16\n",
      "gi|262205423|ref|NR_030217.1| 1.60609e-16\n",
      "gi|301171401|ref|NR_035866.1| 5.6058e-16\n",
      "gi|270133247|ref|NR_032574.1| 5.6058e-16\n",
      "gi|262205309|ref|NR_030193.1| 5.6058e-16\n",
      "gi|270132717|ref|NR_032716.1| 1.95662e-15\n",
      "gi|270132717|ref|NR_032716.1| 0.00171634\n",
      "gi|301171437|ref|NR_035870.1| 6.82927e-15\n",
      "gi|301171437|ref|NR_035870.1| 0.254728\n",
      "gi|270133306|ref|NR_032587.1| 6.82927e-15\n",
      "gi|270133306|ref|NR_032587.1| 9.49283e-07\n",
      "gi|301171428|ref|NR_035869.1| 2.38365e-14\n",
      "gi|301171428|ref|NR_035869.1| 2.71974e-07\n",
      "gi|301171211|ref|NR_035845.1| 2.38365e-14\n",
      "gi|301171153|ref|NR_035838.1| 2.38365e-14\n",
      "gi|301171153|ref|NR_035838.1| 0.0209094\n",
      "gi|301171146|ref|NR_035837.1| 2.38365e-14\n",
      "gi|301171146|ref|NR_035837.1| 0.0209094\n",
      "gi|270133254|ref|NR_032575.1| 2.38365e-14\n",
      "gi|270133254|ref|NR_032575.1| 3.31332e-06\n",
      "gi|262205445|ref|NR_030221.1| 2.38365e-14\n",
      "gi|262205445|ref|NR_030221.1| 0.0209094\n",
      "gi|262205441|ref|NR_030220.1| 2.38365e-14\n",
      "gi|262205441|ref|NR_030220.1| 0.0209094\n",
      "gi|262205325|ref|NR_030197.1| 2.38365e-14\n",
      "gi|262205325|ref|NR_030197.1| 2.71974e-07\n",
      "gi|262205321|ref|NR_030196.1| 2.38365e-14\n",
      "gi|301171236|ref|NR_035848.1| 8.31975e-14\n",
      "gi|301171236|ref|NR_035848.1| 0.000491741\n",
      "gi|262205383|ref|NR_030209.1| 8.31975e-14\n",
      "gi|262205383|ref|NR_030209.1| 0.000491741\n",
      "gi|301171342|ref|NR_035859.1| 2.90388e-13\n",
      "gi|262205265|ref|NR_030183.1| 2.90388e-13\n",
      "gi|301171247|ref|NR_035849.1| 1.01355e-12\n",
      "gi|301171247|ref|NR_035849.1| 2.71974e-07\n",
      "gi|270133223|ref|NR_032569.1| 1.01355e-12\n",
      "gi|262205313|ref|NR_030194.1| 1.01355e-12\n",
      "gi|262205313|ref|NR_030194.1| 2.71974e-07\n",
      "gi|301171225|ref|NR_035847.1| 3.53765e-12\n",
      "gi|301171225|ref|NR_035847.1| 0.000491741\n",
      "gi|270133232|ref|NR_032571.1| 3.53765e-12\n",
      "gi|262205392|ref|NR_030211.1| 3.53765e-12\n",
      "gi|262205392|ref|NR_030211.1| 0.000491741\n",
      "gi|301171300|ref|NR_035855.1| 1.23476e-11\n",
      "gi|301171161|ref|NR_035839.1| 1.23476e-11\n",
      "gi|301171161|ref|NR_035839.1| 0.0729808\n",
      "gi|270133259|ref|NR_032576.1| 1.23476e-11\n",
      "gi|270133259|ref|NR_032576.1| 0.00171634\n",
      "gi|270133218|ref|NR_032568.1| 1.23476e-11\n",
      "gi|270133218|ref|NR_032568.1| 0.00171634\n",
      "gi|270132725|ref|NR_032718.1| 1.23476e-11\n",
      "gi|270132725|ref|NR_032718.1| 0.0729808\n",
      "gi|262205396|ref|NR_030212.1| 1.23476e-11\n",
      "gi|262205396|ref|NR_030212.1| 0.0729808\n",
      "gi|262205294|ref|NR_030189.1| 1.23476e-11\n",
      "gi|301171602|ref|NR_035634.1| 4.30974e-11\n",
      "gi|301171602|ref|NR_035634.1| 0.00599063\n",
      "gi|301171454|ref|NR_035872.1| 4.30974e-11\n",
      "gi|301171333|ref|NR_035858.1| 4.30974e-11\n",
      "gi|301171291|ref|NR_035854.1| 4.30974e-11\n",
      "gi|301171219|ref|NR_035846.1| 4.30974e-11\n",
      "gi|301171219|ref|NR_035846.1| 0.00599063\n",
      "gi|269847477|ref|NR_031696.1| 4.30974e-11\n",
      "gi|262205435|ref|NR_030219.1| 4.30974e-11\n",
      "gi|262205336|ref|NR_030199.1| 4.30974e-11\n",
      "gi|262205336|ref|NR_030199.1| 0.00599063\n",
      "gi|270133264|ref|NR_032577.1| 1.50425e-10\n",
      "gi|269846946|ref|NR_031573.1| 1.50425e-10\n",
      "gi|269846946|ref|NR_031573.1| 0.0209094\n",
      "gi|262205358|ref|NR_030204.1| 1.50425e-10\n",
      "gi|301171206|ref|NR_035844.1| 5.25034e-10\n",
      "gi|301171206|ref|NR_035844.1| 0.00599063\n",
      "gi|301171169|ref|NR_035840.1| 5.25034e-10\n",
      "gi|262205387|ref|NR_030210.1| 5.25034e-10\n",
      "gi|262205387|ref|NR_030210.1| 0.00599063\n",
      "gi|301171381|ref|NR_035863.1| 1.83255e-09\n",
      "gi|270133201|ref|NR_032565.1| 1.83255e-09\n",
      "gi|262205417|ref|NR_030216.1| 1.83255e-09\n",
      "gi|270133214|ref|NR_032567.1| 6.39622e-09\n",
      "gi|262205407|ref|NR_030214.1| 6.39622e-09\n",
      "gi|262205377|ref|NR_030208.1| 6.39622e-09\n",
      "gi|270133311|ref|NR_032588.1| 2.2325e-08\n",
      "gi|270133311|ref|NR_032588.1| 0.0729808\n",
      "gi|270133297|ref|NR_032585.1| 2.2325e-08\n",
      "gi|262205401|ref|NR_030213.1| 2.2325e-08\n",
      "gi|262205401|ref|NR_030213.1| 0.00599063\n",
      "gi|301171372|ref|NR_035862.1| 7.79219e-08\n",
      "gi|301171372|ref|NR_035862.1| 0.00171634\n",
      "gi|262205412|ref|NR_030215.1| 7.79219e-08\n",
      "gi|262205412|ref|NR_030215.1| 0.00171634\n",
      "gi|262205366|ref|NR_030206.1| 7.79219e-08\n",
      "gi|262205366|ref|NR_030206.1| 0.00171634\n",
      "gi|270132721|ref|NR_032717.1| 2.71974e-07\n",
      "gi|270132721|ref|NR_032717.1| 0.889088\n",
      "gi|262205276|ref|NR_030185.1| 9.49283e-07\n",
      "gi|301171363|ref|NR_035861.1| 3.31332e-06\n",
      "gi|270133196|ref|NR_032564.1| 3.31332e-06\n",
      "gi|270133191|ref|NR_032563.1| 3.31332e-06\n",
      "gi|301171387|ref|NR_035864.1| 1.15646e-05\n",
      "gi|262205353|ref|NR_030203.1| 1.15646e-05\n",
      "gi|270133237|ref|NR_032572.1| 0.000140886\n",
      "gi|301171283|ref|NR_035853.1| 0.00171634\n",
      "gi|270133288|ref|NR_032583.1| 0.00171634\n",
      "gi|262205371|ref|NR_030207.1| 0.00171634\n",
      "gi|262205349|ref|NR_030202.1| 0.00171634\n",
      "gi|301171138|ref|NR_035836.1| 0.254728\n",
      "gi|301171409|ref|NR_035867.1| 0.254728\n",
      "gi|301173375|ref|NR_035835.1| 0.254728\n",
      "gi|262205340|ref|NR_030200.1| 0.254728\n",
      "gi|262205285|ref|NR_030187.1| 0.254728\n",
      "gi|262205270|ref|NR_030184.1| 0.254728\n",
      "gi|390332045|ref|XM_776818.2| 0.889088\n",
      "gi|390332043|ref|XM_003723358.1| 0.889088\n",
      "gi|356543101|ref|XM_003539954.1| 0.889088\n",
      "gi|356517317|ref|XM_003527287.1| 0.889088\n",
      "gi|297814701|ref|XM_002875188.1| 0.889088\n",
      "gi|397513516|ref|XM_003827011.1| 3.10322\n"
     ]
    }
   ],
   "source": [
    "# Parse a BLAST XML file\n",
    "\n",
    "from Bio import SearchIO\n",
    "\n",
    "handle = \"searchio-data/blast.xml\"\n",
    "\n",
    "my_result = SearchIO.read(handle = handle, format = \"blast-xml\")\n",
    "print(dir(my_result))\n",
    "\n",
    "print(\"Search {} has {} hits\".format(my_result.id, len(my_result)))\n",
    "\n",
    "print(dir(my_result.hsps[0]))\n",
    "\n",
    "for hsp in my_result.hsps:\n",
    "    print(hsp.hit_id, hsp.evalue) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing and Format Conversion\n",
    "\n",
    "You can also write results files in various file formats. To do this, use the `Bio.SearchIO.write()` function. In the example below, we parse a file in the blast-xml format, and then write its contents to a new file in blast-tab tabular format. \n",
    "\n",
    "```\n",
    "from Bio import SearchIO\n",
    "qresults = SearchIO.parse(handle = 'sample.xml', format = 'blast-xml')\n",
    "SearchIO.write(qresults, handle = 'sample.tab', format = 'blast-tab')\n",
    "```\n",
    "\n",
    "You can also use the `convert()` function to directly convert between file formats without reading/parsing and then writing. The supported formats are below, although not all of the pairs will work for conversion (see [documentation](https://biopython.org/docs/latest/api/Bio.SearchIO.html#Bio.SearchIO.convert) for details).\n",
    "\n",
    "'blast-tab', 'blast-xml', 'blat-psl', 'hmmer3-tab', 'hmmscan3-domtab', 'hmmsearch3-domtab', 'phmmer3-domtab\"\n",
    "\n",
    "The function will return a tuple of four values: the number of QueryResult, Hit, HSP, and HSPFragment objects it writes to the output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_file = \"searchio-data/blast.xml\"\n",
    "out_file = \"searchio-data/blast_tab.tab\"\n",
    "\n",
    "in_format = \"blast-xml\"\n",
    "out_format = \"blast-tab\"\n",
    "\n",
    "SearchIO.convert(in_file=in_file, in_format=in_format, out_file=out_file, out_format=out_format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Entrez\n",
    "\n",
    "The Entrez system is a collection of [NCBI databases](https://www.ncbi.nlm.nih.gov/guide/all/) ranging from literature to sequences, along with a text search tool for exploring them. You can [search online](https://www.ncbi.nlm.nih.gov/search/) via browser, or through the NCBI's E-utilities API. \n",
    "\n",
    "### Entrez Rules and Etiquette\n",
    "\n",
    "### UIDs\n",
    "\n",
    "Every entry in a NCBI database has a UID (unique identifier). This UID will vary depending on the database. For example, PubMed uses PMID while protein records use GI numbers. \n",
    "\n",
    "| Entrez Database    | UID common name | E-utility Database Name |\n",
    "|--------------------|-----------------|-------------------------|\n",
    "| BioProject         | BioProject ID  | bioproject              |\n",
    "| BioSample          | BioSample ID   | biosample               |\n",
    "| Books              | Book ID        | books                   |\n",
    "| Conserved Domains  | PSSM-ID        | cdd                     |\n",
    "| dbGaP              | dbGaP ID       | gap                     |\n",
    "| dbVar              | dbVar ID       | dbvar                   |\n",
    "| Gene               | Gene ID        | gene                    |\n",
    "| Genome             | Genome ID      | genome                  |\n",
    "| GEO Datasets       | GDS ID         | gds                     |\n",
    "| GEO Profiles       | GEO ID         | geoprofiles             |\n",
    "| HomoloGene         | HomoloGene ID  | homologene              |\n",
    "| MeSH               | MeSH ID        | mesh                    |\n",
    "| NCBI C++ Toolkit   | Toolkit ID     | toolkit                 |\n",
    "| NLM Catalog        | NLM Catalog ID | nlmcatalog              |\n",
    "| Nucleotide         | GI number      | nuccore                 |\n",
    "| PopSet             | PopSet ID      | popset                  |\n",
    "| Probe              | Probe ID       | probe                   |\n",
    "| Protein            | GI number      | protein                 |\n",
    "| Protein Clusters   | Protein Cluster ID | proteinclusters      |\n",
    "| PubChem BioAssay   | AID            | pcassay                 |\n",
    "| PubChem Compound   | CID            | pccompound              |\n",
    "| PubChem Substance  | SID            | pcsubstance             |\n",
    "| PubMed             | PMID           | pubmed                  |\n",
    "| PubMed Central     | PMCID          | pmc                     |\n",
    "| SNP                | rs number      | snp                     |\n",
    "| SRA                | SRA ID         | sra                     |\n",
    "| Structure          | MMDB-ID        | structure               |\n",
    "| Taxonomy           | TaxID          | taxonomy                |\n",
    "\n",
    "**Accession.Version vs GI Number**\n",
    "\n",
    "Sequences will have two parallel identifiers given by the NCBI: the GI number and the Accession version. For a full disambiguation, see this [link](https://www.ncbi.nlm.nih.gov/genbank/sequenceids/). \n",
    "\n",
    "### E-Utilities on the Unix Command Line\n",
    "\n",
    "While Biopython gives us nice tools for using the E-utilities withi Python code, they are also available as [command line tools](https://www.ncbi.nlm.nih.gov/books/NBK179288/)\n",
    "\n",
    "### Core Concepts of the E-utilities\n",
    "\n",
    "**What are the E-utilities?**\n",
    "\n",
    "There are nine total E-utilities which perform different tasks with respect to the NCBI databases (from [A General Introduction to the E-utilities](https://www.ncbi.nlm.nih.gov/books/NBK25497/)).\n",
    "\n",
    "1. EInfo (database statistics) eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\n",
    "\n",
    "    Provides the number of records indexed in each field of a given database, the date of the last update of the database, and the available links from the database to other Entrez databases.\n",
    "\n",
    "2. ESearch (text searches) eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\n",
    "\n",
    "    Responds to a text query with the list of matching UIDs in a given database (for later use in ESummary, EFetch or ELink), along with the term translations of the query.\n",
    "\n",
    "3. EPost (UID uploads) eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi\n",
    "\n",
    "    Accepts a list of UIDs from a given database, stores the set on the History Server, and responds with a query key and web environment for the uploaded dataset.\n",
    "\n",
    "4. ESummary (document summary downloads) eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\n",
    "\n",
    "    Responds to a list of UIDs from a given database with the corresponding document summaries.\n",
    "\n",
    "5. EFetch (data record downloads) eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\n",
    "\n",
    "    Responds to a list of UIDs in a given database with the corresponding data records in a specified format.\n",
    "\n",
    "6. ELink (Entrez links) eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\n",
    "\n",
    "    Responds to a list of UIDs in a given database with either a list of related UIDs (and relevancy scores) in the same database or a list of linked UIDs in another Entrez database; checks for the existence of a specified link from a list of one or more UIDs; creates a hyperlink to the primary LinkOut provider for a specific UID and database, or lists LinkOut URLs and attributes for multiple UIDs.\n",
    "\n",
    "7. EGQuery (global query) eutils.ncbi.nlm.nih.gov/entrez/eutils/egquery.fcgi\n",
    "\n",
    "    Responds to a text query with the number of records matching the query in each Entrez database.\n",
    "\n",
    "8. ESpell (spelling suggestions) eutils.ncbi.nlm.nih.gov/entrez/eutils/espell.fcgi\n",
    "\n",
    "    Retrieves spelling suggestions for a text query in a given database.\n",
    "\n",
    "9. ECitMatch (batch citation searching in PubMed) eutils.ncbi.nlm.nih.gov/entrez/eutils/ecitmatch.cgi\n",
    "\n",
    "    Retrieves PubMed IDs (PMIDs) corresponding to a set of input citation strings.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'DbList': ['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']}\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"cwarner@rockefeller.edu\"\n",
    "\n",
    "handle = Entrez.einfo()\n",
    "record = Entrez.read(handle)\n",
    "print(record)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Useful Links and Documentation\n",
    "\n",
    "* [SearchIO](https://biopython.org/docs/latest/api/Bio.SearchIO.html)\n",
    "* [Entrez Help](https://www.ncbi.nlm.nih.gov/books/NBK3837/)\n",
    "\n",
    "## Credits and Inspiration\n",
    "\n",
    "* [Biopython Tutorial and Cookbook](https://biopython.org/DIST/docs/tutorial/Tutorial.html)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
