{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopython 3: SearchIO and Entrez\n",
    "\n",
    "In this workshop, we will take a quick look at SearchIO, Biopython's tool for unifying results from various sequence search tools, before exploring the Entrez API utility from NCBI.\n",
    "\n",
    "## SearchIO\n",
    "\n",
    "Biopython provides the [SearchIO package](https://biopython.org/docs/latest/api/Bio.SearchIO.html) for dealing with outputs from various sequence searching utilities, allowing them to be compared directly. Similar to SeqIO and AlignIO, it has capabilities for parsing, reading, writing, and indexing different search results, as well as converting between file types. \n",
    "\n",
    "### Reading and Parsing\n",
    "\n",
    "Parsing and reading also work the same way as previously, where parse returns an iterator, and read simply returns the first item in the file. Note here that each item is a query, not an individual sequence. \n",
    "\n",
    "Let's start with an example of reading and XML result file from a BLAST search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['_NON_STICKY_ATTRS', '_QueryResult__alt_hit_ids', '_QueryResult__marker', '__annotations__', '__bool__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_description', '_hit_key_function', '_id', '_items', '_transfer_attrs', 'absorb', 'append', 'blast_id', 'description', 'fragments', 'hit_filter', 'hit_keys', 'hit_map', 'hits', 'hsp_filter', 'hsp_map', 'hsps', 'id', 'index', 'items', 'iterhit_keys', 'iterhits', 'iteritems', 'param_evalue_threshold', 'param_filter', 'param_gap_extend', 'param_gap_open', 'param_score_match', 'param_score_mismatch', 'pop', 'program', 'reference', 'seq_len', 'sort', 'stat_db_len', 'stat_db_num', 'stat_eff_space', 'stat_entropy', 'stat_hsp_len', 'stat_kappa', 'stat_lambda', 'target', 'version']\n",
      "Search 42291 has 100 hits\n",
      "['_NON_STICKY_ATTRS', '__annotations__', '__bool__', '__class__', '__contains__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_aln_span_get', '_get_coords', '_hit_end_get', '_hit_inter_ranges_get', '_hit_inter_spans_get', '_hit_range_get', '_hit_span_get', '_hit_start_get', '_inter_ranges_get', '_inter_spans_get', '_items', '_query_end_get', '_query_inter_ranges_get', '_query_inter_spans_get', '_query_range_get', '_query_span_get', '_query_start_get', '_str_hsp_header', '_transfer_attrs', '_validate_fragment', 'aln', 'aln_all', 'aln_annotation', 'aln_annotation_all', 'aln_span', 'bitscore', 'bitscore_raw', 'evalue', 'fragment', 'fragments', 'gap_num', 'hit', 'hit_all', 'hit_description', 'hit_end', 'hit_end_all', 'hit_features', 'hit_features_all', 'hit_frame', 'hit_frame_all', 'hit_id', 'hit_inter_ranges', 'hit_inter_spans', 'hit_range', 'hit_range_all', 'hit_span', 'hit_span_all', 'hit_start', 'hit_start_all', 'hit_strand', 'hit_strand_all', 'ident_num', 'is_fragmented', 'molecule_type', 'output_index', 'pos_num', 'query', 'query_all', 'query_description', 'query_end', 'query_end_all', 'query_features', 'query_features_all', 'query_frame', 'query_frame_all', 'query_id', 'query_inter_ranges', 'query_inter_spans', 'query_range', 'query_range_all', 'query_span', 'query_span_all', 'query_start', 'query_start_all', 'query_strand', 'query_strand_all']\n",
      "gi|262205317|ref|NR_030195.1| 4.91307e-23\n",
      "gi|301171311|ref|NR_035856.1| 1.71483e-22\n",
      "gi|270133242|ref|NR_032573.1| 2.54503e-20\n",
      "gi|301171322|ref|NR_035857.1| 8.88303e-20\n",
      "gi|301171322|ref|NR_035857.1| 3.31332e-06\n",
      "gi|301171267|ref|NR_035851.1| 8.88303e-20\n",
      "gi|262205330|ref|NR_030198.1| 8.88303e-20\n",
      "gi|262205330|ref|NR_030198.1| 3.31332e-06\n",
      "gi|262205302|ref|NR_030191.1| 8.88303e-20\n",
      "gi|301171259|ref|NR_035850.1| 3.10048e-19\n",
      "gi|262205451|ref|NR_030222.1| 3.10048e-19\n",
      "gi|301171447|ref|NR_035871.1| 3.77716e-18\n",
      "gi|301171447|ref|NR_035871.1| 0.000140886\n",
      "gi|301171276|ref|NR_035852.1| 3.77716e-18\n",
      "gi|262205290|ref|NR_030188.1| 3.77716e-18\n",
      "gi|301171354|ref|NR_035860.1| 1.31836e-17\n",
      "gi|262205281|ref|NR_030186.1| 1.31836e-17\n",
      "gi|262205298|ref|NR_030190.1| 4.60152e-17\n",
      "gi|262205298|ref|NR_030190.1| 0.000140886\n",
      "gi|301171394|ref|NR_035865.1| 1.60609e-16\n",
      "gi|262205429|ref|NR_030218.1| 1.60609e-16\n",
      "gi|262205423|ref|NR_030217.1| 1.60609e-16\n",
      "gi|301171401|ref|NR_035866.1| 5.6058e-16\n",
      "gi|270133247|ref|NR_032574.1| 5.6058e-16\n",
      "gi|262205309|ref|NR_030193.1| 5.6058e-16\n",
      "gi|270132717|ref|NR_032716.1| 1.95662e-15\n",
      "gi|270132717|ref|NR_032716.1| 0.00171634\n",
      "gi|301171437|ref|NR_035870.1| 6.82927e-15\n",
      "gi|301171437|ref|NR_035870.1| 0.254728\n",
      "gi|270133306|ref|NR_032587.1| 6.82927e-15\n",
      "gi|270133306|ref|NR_032587.1| 9.49283e-07\n",
      "gi|301171428|ref|NR_035869.1| 2.38365e-14\n",
      "gi|301171428|ref|NR_035869.1| 2.71974e-07\n",
      "gi|301171211|ref|NR_035845.1| 2.38365e-14\n",
      "gi|301171153|ref|NR_035838.1| 2.38365e-14\n",
      "gi|301171153|ref|NR_035838.1| 0.0209094\n",
      "gi|301171146|ref|NR_035837.1| 2.38365e-14\n",
      "gi|301171146|ref|NR_035837.1| 0.0209094\n",
      "gi|270133254|ref|NR_032575.1| 2.38365e-14\n",
      "gi|270133254|ref|NR_032575.1| 3.31332e-06\n",
      "gi|262205445|ref|NR_030221.1| 2.38365e-14\n",
      "gi|262205445|ref|NR_030221.1| 0.0209094\n",
      "gi|262205441|ref|NR_030220.1| 2.38365e-14\n",
      "gi|262205441|ref|NR_030220.1| 0.0209094\n",
      "gi|262205325|ref|NR_030197.1| 2.38365e-14\n",
      "gi|262205325|ref|NR_030197.1| 2.71974e-07\n",
      "gi|262205321|ref|NR_030196.1| 2.38365e-14\n",
      "gi|301171236|ref|NR_035848.1| 8.31975e-14\n",
      "gi|301171236|ref|NR_035848.1| 0.000491741\n",
      "gi|262205383|ref|NR_030209.1| 8.31975e-14\n",
      "gi|262205383|ref|NR_030209.1| 0.000491741\n",
      "gi|301171342|ref|NR_035859.1| 2.90388e-13\n",
      "gi|262205265|ref|NR_030183.1| 2.90388e-13\n",
      "gi|301171247|ref|NR_035849.1| 1.01355e-12\n",
      "gi|301171247|ref|NR_035849.1| 2.71974e-07\n",
      "gi|270133223|ref|NR_032569.1| 1.01355e-12\n",
      "gi|262205313|ref|NR_030194.1| 1.01355e-12\n",
      "gi|262205313|ref|NR_030194.1| 2.71974e-07\n",
      "gi|301171225|ref|NR_035847.1| 3.53765e-12\n",
      "gi|301171225|ref|NR_035847.1| 0.000491741\n",
      "gi|270133232|ref|NR_032571.1| 3.53765e-12\n",
      "gi|262205392|ref|NR_030211.1| 3.53765e-12\n",
      "gi|262205392|ref|NR_030211.1| 0.000491741\n",
      "gi|301171300|ref|NR_035855.1| 1.23476e-11\n",
      "gi|301171161|ref|NR_035839.1| 1.23476e-11\n",
      "gi|301171161|ref|NR_035839.1| 0.0729808\n",
      "gi|270133259|ref|NR_032576.1| 1.23476e-11\n",
      "gi|270133259|ref|NR_032576.1| 0.00171634\n",
      "gi|270133218|ref|NR_032568.1| 1.23476e-11\n",
      "gi|270133218|ref|NR_032568.1| 0.00171634\n",
      "gi|270132725|ref|NR_032718.1| 1.23476e-11\n",
      "gi|270132725|ref|NR_032718.1| 0.0729808\n",
      "gi|262205396|ref|NR_030212.1| 1.23476e-11\n",
      "gi|262205396|ref|NR_030212.1| 0.0729808\n",
      "gi|262205294|ref|NR_030189.1| 1.23476e-11\n",
      "gi|301171602|ref|NR_035634.1| 4.30974e-11\n",
      "gi|301171602|ref|NR_035634.1| 0.00599063\n",
      "gi|301171454|ref|NR_035872.1| 4.30974e-11\n",
      "gi|301171333|ref|NR_035858.1| 4.30974e-11\n",
      "gi|301171291|ref|NR_035854.1| 4.30974e-11\n",
      "gi|301171219|ref|NR_035846.1| 4.30974e-11\n",
      "gi|301171219|ref|NR_035846.1| 0.00599063\n",
      "gi|269847477|ref|NR_031696.1| 4.30974e-11\n",
      "gi|262205435|ref|NR_030219.1| 4.30974e-11\n",
      "gi|262205336|ref|NR_030199.1| 4.30974e-11\n",
      "gi|262205336|ref|NR_030199.1| 0.00599063\n",
      "gi|270133264|ref|NR_032577.1| 1.50425e-10\n",
      "gi|269846946|ref|NR_031573.1| 1.50425e-10\n",
      "gi|269846946|ref|NR_031573.1| 0.0209094\n",
      "gi|262205358|ref|NR_030204.1| 1.50425e-10\n",
      "gi|301171206|ref|NR_035844.1| 5.25034e-10\n",
      "gi|301171206|ref|NR_035844.1| 0.00599063\n",
      "gi|301171169|ref|NR_035840.1| 5.25034e-10\n",
      "gi|262205387|ref|NR_030210.1| 5.25034e-10\n",
      "gi|262205387|ref|NR_030210.1| 0.00599063\n",
      "gi|301171381|ref|NR_035863.1| 1.83255e-09\n",
      "gi|270133201|ref|NR_032565.1| 1.83255e-09\n",
      "gi|262205417|ref|NR_030216.1| 1.83255e-09\n",
      "gi|270133214|ref|NR_032567.1| 6.39622e-09\n",
      "gi|262205407|ref|NR_030214.1| 6.39622e-09\n",
      "gi|262205377|ref|NR_030208.1| 6.39622e-09\n",
      "gi|270133311|ref|NR_032588.1| 2.2325e-08\n",
      "gi|270133311|ref|NR_032588.1| 0.0729808\n",
      "gi|270133297|ref|NR_032585.1| 2.2325e-08\n",
      "gi|262205401|ref|NR_030213.1| 2.2325e-08\n",
      "gi|262205401|ref|NR_030213.1| 0.00599063\n",
      "gi|301171372|ref|NR_035862.1| 7.79219e-08\n",
      "gi|301171372|ref|NR_035862.1| 0.00171634\n",
      "gi|262205412|ref|NR_030215.1| 7.79219e-08\n",
      "gi|262205412|ref|NR_030215.1| 0.00171634\n",
      "gi|262205366|ref|NR_030206.1| 7.79219e-08\n",
      "gi|262205366|ref|NR_030206.1| 0.00171634\n",
      "gi|270132721|ref|NR_032717.1| 2.71974e-07\n",
      "gi|270132721|ref|NR_032717.1| 0.889088\n",
      "gi|262205276|ref|NR_030185.1| 9.49283e-07\n",
      "gi|301171363|ref|NR_035861.1| 3.31332e-06\n",
      "gi|270133196|ref|NR_032564.1| 3.31332e-06\n",
      "gi|270133191|ref|NR_032563.1| 3.31332e-06\n",
      "gi|301171387|ref|NR_035864.1| 1.15646e-05\n",
      "gi|262205353|ref|NR_030203.1| 1.15646e-05\n",
      "gi|270133237|ref|NR_032572.1| 0.000140886\n",
      "gi|301171283|ref|NR_035853.1| 0.00171634\n",
      "gi|270133288|ref|NR_032583.1| 0.00171634\n",
      "gi|262205371|ref|NR_030207.1| 0.00171634\n",
      "gi|262205349|ref|NR_030202.1| 0.00171634\n",
      "gi|301171138|ref|NR_035836.1| 0.254728\n",
      "gi|301171409|ref|NR_035867.1| 0.254728\n",
      "gi|301173375|ref|NR_035835.1| 0.254728\n",
      "gi|262205340|ref|NR_030200.1| 0.254728\n",
      "gi|262205285|ref|NR_030187.1| 0.254728\n",
      "gi|262205270|ref|NR_030184.1| 0.254728\n",
      "gi|390332045|ref|XM_776818.2| 0.889088\n",
      "gi|390332043|ref|XM_003723358.1| 0.889088\n",
      "gi|356543101|ref|XM_003539954.1| 0.889088\n",
      "gi|356517317|ref|XM_003527287.1| 0.889088\n",
      "gi|297814701|ref|XM_002875188.1| 0.889088\n",
      "gi|397513516|ref|XM_003827011.1| 3.10322\n"
     ]
    }
   ],
   "source": [
    "# Parse a BLAST XML file\n",
    "\n",
    "from Bio import SearchIO\n",
    "\n",
    "handle = \"searchio-data/blast.xml\"\n",
    "\n",
    "my_result = SearchIO.read(handle = handle, format = \"blast-xml\")\n",
    "print(dir(my_result))\n",
    "\n",
    "print(\"Search {} has {} hits\".format(my_result.id, len(my_result)))\n",
    "\n",
    "print(dir(my_result.hsps[0]))\n",
    "\n",
    "for hsp in my_result.hsps:\n",
    "    print(hsp.hit_id, hsp.evalue) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing and Format Conversion\n",
    "\n",
    "You can also write results files in various file formats. To do this, use the `Bio.SearchIO.write()` function. In the example below, we parse a file in the blast-xml format, and then write its contents to a new file in blast-tab tabular format. \n",
    "\n",
    "```\n",
    "from Bio import SearchIO\n",
    "qresults = SearchIO.parse(handle = 'sample.xml', format = 'blast-xml')\n",
    "SearchIO.write(qresults, handle = 'sample.tab', format = 'blast-tab')\n",
    "```\n",
    "\n",
    "You can also use the `convert()` function to directly convert between file formats without reading/parsing and then writing. The supported formats are below, although not all of the pairs will work for conversion (see [documentation](https://biopython.org/docs/latest/api/Bio.SearchIO.html#Bio.SearchIO.convert) for details).\n",
    "\n",
    "'blast-tab', 'blast-xml', 'blat-psl', 'hmmer3-tab', 'hmmscan3-domtab', 'hmmsearch3-domtab', 'phmmer3-domtab\"\n",
    "\n",
    "The function will return a tuple of four values: the number of QueryResult, Hit, HSP, and HSPFragment objects it writes to the output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_file = \"searchio-data/blast.xml\"\n",
    "out_file = \"searchio-data/blast_tab.tab\"\n",
    "\n",
    "in_format = \"blast-xml\"\n",
    "out_format = \"blast-tab\"\n",
    "\n",
    "SearchIO.convert(in_file=in_file, in_format=in_format, out_file=out_file, out_format=out_format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Entrez\n",
    "\n",
    "The Entrez system is a collection of [NCBI databases](https://www.ncbi.nlm.nih.gov/guide/all/) ranging from literature to sequences, along with a text search tool for exploring them. You can [search online](https://www.ncbi.nlm.nih.gov/search/) via browser, or through the NCBI's E-utilities API. \n",
    "\n",
    "### Entrez Rules and Etiquette\n",
    "\n",
    "### UIDs\n",
    "\n",
    "Every entry in a NCBI database has a UID (unique identifier). This UID will vary depending on the database. For example, PubMed uses PMID while protein records use GI numbers. \n",
    "\n",
    "| Entrez Database    | UID common name | E-utility Database Name |\n",
    "|--------------------|-----------------|-------------------------|\n",
    "| BioProject         | BioProject ID  | bioproject              |\n",
    "| BioSample          | BioSample ID   | biosample               |\n",
    "| Books              | Book ID        | books                   |\n",
    "| Conserved Domains  | PSSM-ID        | cdd                     |\n",
    "| dbGaP              | dbGaP ID       | gap                     |\n",
    "| dbVar              | dbVar ID       | dbvar                   |\n",
    "| Gene               | Gene ID        | gene                    |\n",
    "| Genome             | Genome ID      | genome                  |\n",
    "| GEO Datasets       | GDS ID         | gds                     |\n",
    "| GEO Profiles       | GEO ID         | geoprofiles             |\n",
    "| HomoloGene         | HomoloGene ID  | homologene              |\n",
    "| MeSH               | MeSH ID        | mesh                    |\n",
    "| NCBI C++ Toolkit   | Toolkit ID     | toolkit                 |\n",
    "| NLM Catalog        | NLM Catalog ID | nlmcatalog              |\n",
    "| Nucleotide         | GI number      | nuccore                 |\n",
    "| PopSet             | PopSet ID      | popset                  |\n",
    "| Probe              | Probe ID       | probe                   |\n",
    "| Protein            | GI number      | protein                 |\n",
    "| Protein Clusters   | Protein Cluster ID | proteinclusters      |\n",
    "| PubChem BioAssay   | AID            | pcassay                 |\n",
    "| PubChem Compound   | CID            | pccompound              |\n",
    "| PubChem Substance  | SID            | pcsubstance             |\n",
    "| PubMed             | PMID           | pubmed                  |\n",
    "| PubMed Central     | PMCID          | pmc                     |\n",
    "| SNP                | rs number      | snp                     |\n",
    "| SRA                | SRA ID         | sra                     |\n",
    "| Structure          | MMDB-ID        | structure               |\n",
    "| Taxonomy           | TaxID          | taxonomy                |\n",
    "\n",
    "**Accession.Version vs GI Number**\n",
    "\n",
    "Sequences will have two parallel identifiers given by the NCBI: the GI number and the Accession version. For a full disambiguation, see this [link](https://www.ncbi.nlm.nih.gov/genbank/sequenceids/). \n",
    "\n",
    "### E-Utilities on the Unix Command Line\n",
    "\n",
    "While Biopython gives us nice tools for using the E-utilities withi Python code, they are also available as [command line tools](https://www.ncbi.nlm.nih.gov/books/NBK179288/)\n",
    "\n",
    "### Core Concepts of the E-utilities\n",
    "\n",
    "**What are the E-utilities?**\n",
    "\n",
    "There are nine total E-utilities which perform different tasks with respect to the NCBI databases (from [A General Introduction to the E-utilities](https://www.ncbi.nlm.nih.gov/books/NBK25497/)).\n",
    "\n",
    "1. EInfo (database statistics) eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi\n",
    "\n",
    "    Provides the number of records indexed in each field of a given database, the date of the last update of the database, and the available links from the database to other Entrez databases.\n",
    "\n",
    "2. ESearch (text searches) eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi\n",
    "\n",
    "    Responds to a text query with the list of matching UIDs in a given database (for later use in ESummary, EFetch or ELink), along with the term translations of the query.\n",
    "\n",
    "3. EPost (UID uploads) eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi\n",
    "\n",
    "    Accepts a list of UIDs from a given database, stores the set on the History Server, and responds with a query key and web environment for the uploaded dataset.\n",
    "\n",
    "4. ESummary (document summary downloads) eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi\n",
    "\n",
    "    Responds to a list of UIDs from a given database with the corresponding document summaries.\n",
    "\n",
    "5. EFetch (data record downloads) eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\n",
    "\n",
    "    Responds to a list of UIDs in a given database with the corresponding data records in a specified format.\n",
    "\n",
    "6. ELink (Entrez links) eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi\n",
    "\n",
    "    Responds to a list of UIDs in a given database with either a list of related UIDs (and relevancy scores) in the same database or a list of linked UIDs in another Entrez database; checks for the existence of a specified link from a list of one or more UIDs; creates a hyperlink to the primary LinkOut provider for a specific UID and database, or lists LinkOut URLs and attributes for multiple UIDs.\n",
    "\n",
    "7. EGQuery (global query) eutils.ncbi.nlm.nih.gov/entrez/eutils/egquery.fcgi\n",
    "\n",
    "    Responds to a text query with the number of records matching the query in each Entrez database.\n",
    "\n",
    "8. ESpell (spelling suggestions) eutils.ncbi.nlm.nih.gov/entrez/eutils/espell.fcgi\n",
    "\n",
    "    Retrieves spelling suggestions for a text query in a given database.\n",
    "\n",
    "9. ECitMatch (batch citation searching in PubMed) eutils.ncbi.nlm.nih.gov/entrez/eutils/ecitmatch.cgi\n",
    "\n",
    "    Retrieves PubMed IDs (PMIDs) corresponding to a set of input citation strings.\n",
    "\n",
    "## Bio.Entrez Module\n",
    "\n",
    "Biopython gives us a module that allows us to perform these functions inside of a Python script. What it essentially does is to take our arguments, generate the Entrez E-utility URL, send to the NCBI server, and deal with the returned information. The documentation can be found [here](https://biopython.org/docs/latest/api/Bio.Entrez.html). As well as the 9 core functions, the module provides functions for reading and parsing the results.\n",
    "\n",
    "### An Example with EInfo\n",
    "\n",
    "Here we will use the first function, EInfo, to get information on some NCBI databases. The basic routine is as follows:\n",
    "\n",
    "0. Give Entrez your email address.\n",
    "1. Define your handle as the output of the `Entrez.einfo()` function.\n",
    "2. Create a record by \"reading\" the handle using the `Entrez.read()` function. This record will be a nest of Python dictionaries - explore the list by looking at the keys.\n",
    "3. Extract the information you need by calling the specific keys as needed.\n",
    "\n",
    "When given no arguments, `einfo()` returns a list of all valid Entrez databases. You can also specify a database by giving it a `db` (database) variable. Then, it will return the information related to that specific database. There are a couple of other inputs for specifying the output format (JSON is available as well as the default XML) and a version input (which is for the very interested user only - check the documentation for details).\n",
    "\n",
    "First, let's get all of the database information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['DbList'])\n",
      "{'DbList': ['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']}\n",
      "40\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"cwarner@rockefeller.edu\"\n",
    "handle = Entrez.einfo() # Get handle with information\n",
    "record = Entrez.read(handle) # Use the Bio parser to turn it into a Python object (dictionary)\n",
    "print(record.keys())\n",
    "print(record)\n",
    "print(len(record['DbList']))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we can find out more information about the SRA (Sequence Read Archive) database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['DbInfo'])\n",
      "{'DbName': 'gap', 'MenuName': 'dbGaP', 'Description': 'dbGaP Data', 'DbBuild': 'Build230522-0335m.1', 'Count': '363717', 'LastUpdate': '2023/05/22 04:11', 'FieldList': [{'Name': 'ALL', 'FullName': 'All Fields', 'Description': 'All terms from all searchable fields', 'TermCount': '2183077', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'N', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'UID', 'FullName': 'UID', 'Description': 'Unique number assigned to publication', 'TermCount': '0', 'IsDate': 'N', 'IsNumerical': 'Y', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'Y'}, {'Name': 'FILT', 'FullName': 'Filter', 'Description': 'Limits the records', 'TermCount': '23', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DISC', 'FullName': 'Discriminator', 'Description': 'Discriminator', 'TermCount': '10', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANCE', 'FullName': 'Ancestor', 'Description': 'Ancestor', 'TermCount': '2839', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'BELO', 'FullName': 'Belongs To', 'Description': 'Belongs To', 'TermCount': '2836', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ATTR', 'FullName': 'Attribution', 'Description': 'Attribution', 'TermCount': '16170', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'RTST', 'FullName': 'Is Root Study', 'Description': 'Is Root Study', 'TermCount': '7', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'TLST', 'FullName': 'Is Top-Level Study', 'Description': 'Is Top-Level Study', 'TermCount': '7', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'STID', 'FullName': 'Study ID', 'Description': 'Study ID', 'TermCount': '2836', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'STNM', 'FullName': 'Study Name', 'Description': 'Study Name', 'TermCount': '6491', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DIS', 'FullName': 'Disease', 'Description': 'Disease', 'TermCount': '2646', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'PROJ', 'FullName': 'Project', 'Description': 'Project', 'TermCount': '477', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'GENO', 'FullName': 'Genotype Platform', 'Description': 'Genotype Platform', 'TermCount': '1423', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'SRA', 'FullName': 'Study Has SRA Components', 'Description': 'Study Has SRA components', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'STUD', 'FullName': 'Study', 'Description': 'Study', 'TermCount': '85745', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASV', 'FullName': 'Has Variable', 'Description': 'Has Variable', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VRID', 'FullName': 'Variable ID', 'Description': 'Variable ID', 'TermCount': '680333', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VRNM', 'FullName': 'Variable Name', 'Description': 'Variable Name', 'TermCount': '223739', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VRDS', 'FullName': 'Variable Description', 'Description': 'Variable Description', 'TermCount': '163304', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VAR', 'FullName': 'Variable', 'Description': 'Variable', 'TermCount': '1494355', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASD', 'FullName': 'Has Document', 'Description': 'Has Document', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DCID', 'FullName': 'Document ID', 'Description': 'Document ID', 'TermCount': '13281', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DCNM', 'FullName': 'Document Name', 'Description': 'Document Name', 'TermCount': '10738', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DOC', 'FullName': 'Document', 'Description': 'Document', 'TermCount': '764966', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DOCP', 'FullName': 'Document Part', 'Description': 'Document Part', 'TermCount': '123556', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASA', 'FullName': 'Has Analysis', 'Description': 'Has Analysis', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANID', 'FullName': 'Analysis ID', 'Description': 'Analysis ID', 'TermCount': '8952', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANNM', 'FullName': 'Analysis Name', 'Description': 'Analysis Name', 'TermCount': '4093', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANLS', 'FullName': 'Analysis', 'Description': 'Analysis', 'TermCount': '26721', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HAST', 'FullName': 'Has Dataset', 'Description': 'Has Dataset', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DSID', 'FullName': 'Dataset ID', 'Description': 'Dataset ID', 'TermCount': '23964', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DSNM', 'FullName': 'Dataset Name', 'Description': 'Dataset Name', 'TermCount': '11508', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DS', 'FullName': 'Dataset', 'Description': 'Dataset', 'TermCount': '65928', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'PX', 'FullName': 'PhenX', 'Description': 'PhenX', 'TermCount': '3637', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASP', 'FullName': 'Has PhenX Mapping', 'Description': 'Has PhenX Mapping', 'TermCount': '3', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ARCH', 'FullName': 'Study Archive', 'Description': 'Study Archive', 'TermCount': '1', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'MOLE', 'FullName': 'Molecular Data Type', 'Description': 'Molecular Data Type', 'TermCount': '79', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'PRPH', 'FullName': 'Primary Phenotype in Study', 'Description': 'Primary Phenotype in Study', 'TermCount': '585', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DAC', 'FullName': 'Data Access Committee', 'Description': 'Data Access Committee', 'TermCount': '22', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DESN', 'FullName': 'Study Design', 'Description': 'Study Design', 'TermCount': '15', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'CDER', 'FullName': 'Common Data Element Resource', 'Description': 'Common Data Element Resource (e.g. LOINC, PhenX)', 'TermCount': '3', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'CDET', 'FullName': 'Common Data Element Term', 'Description': 'Common Data Element Term (e.g. PhenX accession)', 'TermCount': '2074', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}], 'LinkList': [{'Name': 'gap_bioproject', 'Menu': 'BioProject Links', 'Description': 'BioProject Links', 'DbTo': 'bioproject'}, {'Name': 'gap_biosample_all', 'Menu': 'BioSample Links', 'Description': 'All related BioSamples', 'DbTo': 'biosample'}, {'Name': 'gap_dbvar', 'Menu': 'dbVar', 'Description': 'Link from dbGaP to dbVar', 'DbTo': 'dbvar'}, {'Name': 'gap_gds', 'Menu': 'GEO DataSets', 'Description': 'GEO DataSet links', 'DbTo': 'gds'}, {'Name': 'gap_gene', 'Menu': 'Gene Links', 'Description': 'Gene Links', 'DbTo': 'gene'}, {'Name': 'gap_medgen', 'Menu': 'MedGen', 'Description': 'MedGen Links', 'DbTo': 'medgen'}, {'Name': 'gap_mesh', 'Menu': 'MeSH', 'Description': 'MeSH links', 'DbTo': 'mesh'}, {'Name': 'gap_pmc', 'Menu': 'PMC Links', 'Description': 'PMC Links', 'DbTo': 'pmc'}, {'Name': 'gap_pubmed', 'Menu': 'PubMed Links', 'Description': 'Related PubMed article', 'DbTo': 'pubmed'}, {'Name': 'gap_snp', 'Menu': 'SNP Links', 'Description': 'SNP Links', 'DbTo': 'snp'}, {'Name': 'gap_sra_all', 'Menu': 'SRA Links', 'Description': 'All related SRA records', 'DbTo': 'sra'}]}\n",
      "[{'Name': 'ALL', 'FullName': 'All Fields', 'Description': 'All terms from all searchable fields', 'TermCount': '2183077', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'N', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'UID', 'FullName': 'UID', 'Description': 'Unique number assigned to publication', 'TermCount': '0', 'IsDate': 'N', 'IsNumerical': 'Y', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'Y'}, {'Name': 'FILT', 'FullName': 'Filter', 'Description': 'Limits the records', 'TermCount': '23', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DISC', 'FullName': 'Discriminator', 'Description': 'Discriminator', 'TermCount': '10', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANCE', 'FullName': 'Ancestor', 'Description': 'Ancestor', 'TermCount': '2839', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'BELO', 'FullName': 'Belongs To', 'Description': 'Belongs To', 'TermCount': '2836', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ATTR', 'FullName': 'Attribution', 'Description': 'Attribution', 'TermCount': '16170', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'RTST', 'FullName': 'Is Root Study', 'Description': 'Is Root Study', 'TermCount': '7', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'TLST', 'FullName': 'Is Top-Level Study', 'Description': 'Is Top-Level Study', 'TermCount': '7', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'STID', 'FullName': 'Study ID', 'Description': 'Study ID', 'TermCount': '2836', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'STNM', 'FullName': 'Study Name', 'Description': 'Study Name', 'TermCount': '6491', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DIS', 'FullName': 'Disease', 'Description': 'Disease', 'TermCount': '2646', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'PROJ', 'FullName': 'Project', 'Description': 'Project', 'TermCount': '477', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'GENO', 'FullName': 'Genotype Platform', 'Description': 'Genotype Platform', 'TermCount': '1423', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'SRA', 'FullName': 'Study Has SRA Components', 'Description': 'Study Has SRA components', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'STUD', 'FullName': 'Study', 'Description': 'Study', 'TermCount': '85745', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASV', 'FullName': 'Has Variable', 'Description': 'Has Variable', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VRID', 'FullName': 'Variable ID', 'Description': 'Variable ID', 'TermCount': '680333', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VRNM', 'FullName': 'Variable Name', 'Description': 'Variable Name', 'TermCount': '223739', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VRDS', 'FullName': 'Variable Description', 'Description': 'Variable Description', 'TermCount': '163304', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'VAR', 'FullName': 'Variable', 'Description': 'Variable', 'TermCount': '1494355', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASD', 'FullName': 'Has Document', 'Description': 'Has Document', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DCID', 'FullName': 'Document ID', 'Description': 'Document ID', 'TermCount': '13281', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DCNM', 'FullName': 'Document Name', 'Description': 'Document Name', 'TermCount': '10738', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DOC', 'FullName': 'Document', 'Description': 'Document', 'TermCount': '764966', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DOCP', 'FullName': 'Document Part', 'Description': 'Document Part', 'TermCount': '123556', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASA', 'FullName': 'Has Analysis', 'Description': 'Has Analysis', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANID', 'FullName': 'Analysis ID', 'Description': 'Analysis ID', 'TermCount': '8952', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANNM', 'FullName': 'Analysis Name', 'Description': 'Analysis Name', 'TermCount': '4093', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ANLS', 'FullName': 'Analysis', 'Description': 'Analysis', 'TermCount': '26721', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HAST', 'FullName': 'Has Dataset', 'Description': 'Has Dataset', 'TermCount': '4', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DSID', 'FullName': 'Dataset ID', 'Description': 'Dataset ID', 'TermCount': '23964', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DSNM', 'FullName': 'Dataset Name', 'Description': 'Dataset Name', 'TermCount': '11508', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DS', 'FullName': 'Dataset', 'Description': 'Dataset', 'TermCount': '65928', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'PX', 'FullName': 'PhenX', 'Description': 'PhenX', 'TermCount': '3637', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'HASP', 'FullName': 'Has PhenX Mapping', 'Description': 'Has PhenX Mapping', 'TermCount': '3', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'ARCH', 'FullName': 'Study Archive', 'Description': 'Study Archive', 'TermCount': '1', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'MOLE', 'FullName': 'Molecular Data Type', 'Description': 'Molecular Data Type', 'TermCount': '79', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'PRPH', 'FullName': 'Primary Phenotype in Study', 'Description': 'Primary Phenotype in Study', 'TermCount': '585', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DAC', 'FullName': 'Data Access Committee', 'Description': 'Data Access Committee', 'TermCount': '22', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'DESN', 'FullName': 'Study Design', 'Description': 'Study Design', 'TermCount': '15', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'CDER', 'FullName': 'Common Data Element Resource', 'Description': 'Common Data Element Resource (e.g. LOINC, PhenX)', 'TermCount': '3', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}, {'Name': 'CDET', 'FullName': 'Common Data Element Term', 'Description': 'Common Data Element Term (e.g. PhenX accession)', 'TermCount': '2074', 'IsDate': 'N', 'IsNumerical': 'N', 'SingleToken': 'Y', 'Hierarchy': 'N', 'IsHidden': 'N'}]\n"
     ]
    }
   ],
   "source": [
    "Entrez.email = \"cwarner@rockefeller.edu\"\n",
    "handle = Entrez.einfo(db = \"sra\") # Get handle with information\n",
    "record = Entrez.read(handle) # Change into Python object\n",
    "print(record.keys()) # Print the first \"dictionary layer\"\n",
    "print(record[\"DbInfo\"])\n",
    "print(record[\"DbInfo\"][\"FieldList\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## An Example with Genbank Data\n",
    "\n",
    "Here we will use several of the E-utilities to search for genomic data of the RPL16 gene of a prickly pear (opuntia) in the [NCBI Nucleotide Database](https://www.ncbi.nlm.nih.gov/nucleotide/). This will occur in several stages:\n",
    "\n",
    "1. Use EGQuery to find the number of hits for your search terms in the \"nuccore\" database.\n",
    "2. Use ESearch to get the UIDs searching in \"nuccore\" with the desired search terms\n",
    "3. Use EFetch to get the data associated  with those UIDs\n",
    "4. Parse this sequence data and write it to a .gb file\n",
    "5. Check your work by parsing the .gb file\n",
    "\n",
    "Let's begin by finding the number of results in the \"nuccore\" database.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['Term', 'eGQueryResult'])\n",
      "[{'DbName': 'pubmed', 'MenuName': 'PubMed', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'pmc', 'MenuName': 'PubMed Central', 'Count': '21', 'Status': 'Ok'}, {'DbName': 'mesh', 'MenuName': 'MeSH', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'books', 'MenuName': 'Books', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'pubmedhealth', 'MenuName': 'PubMed Health', 'Count': 'Error', 'Status': 'Database Error'}, {'DbName': 'omim', 'MenuName': 'OMIM', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'ncbisearch', 'MenuName': 'Site Search', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'nuccore', 'MenuName': 'Nucleotide', 'Count': '115', 'Status': 'Ok'}, {'DbName': 'nucgss', 'MenuName': 'GSS', 'Count': '0', 'Status': 'Ok'}, {'DbName': 'nucest', 'MenuName': 'EST', 'Count': '0', 'Status': 'Ok'}, {'DbName': 'protein', 'MenuName': 'Protein', 'Count': '90', 'Status': 'Ok'}, {'DbName': 'genome', 'MenuName': 'Genome', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'structure', 'MenuName': 'Structure', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'taxonomy', 'MenuName': 'Taxonomy', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'snp', 'MenuName': 'SNP', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'dbvar', 'MenuName': 'dbVar', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'gene', 'MenuName': 'Gene', 'Count': '31', 'Status': 'Ok'}, {'DbName': 'sra', 'MenuName': 'SRA', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'biosystems', 'MenuName': 'BioSystems', 'Count': 'Error', 'Status': 'Database Error'}, {'DbName': 'unigene', 'MenuName': 'UniGene', 'Count': '0', 'Status': 'Ok'}, {'DbName': 'cdd', 'MenuName': 'Conserved Domains', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'clone', 'MenuName': 'Clone', 'Count': '0', 'Status': 'Ok'}, {'DbName': 'popset', 'MenuName': 'PopSet', 'Count': '5', 'Status': 'Ok'}, {'DbName': 'geoprofiles', 'MenuName': 'GEO Profiles', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'gds', 'MenuName': 'GEO DataSets', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'homologene', 'MenuName': 'HomoloGene', 'Count': 'Error', 'Status': 'Database Error'}, {'DbName': 'pccompound', 'MenuName': 'PubChem Compound', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'pcsubstance', 'MenuName': 'PubChem Substance', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'pcassay', 'MenuName': 'PubChem BioAssay', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'nlmcatalog', 'MenuName': 'NLM Catalog', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'probe', 'MenuName': 'Probe', 'Count': '0', 'Status': 'Ok'}, {'DbName': 'gap', 'MenuName': 'dbGaP', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'proteinclusters', 'MenuName': 'Protein Clusters', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'bioproject', 'MenuName': 'BioProject', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'biosample', 'MenuName': 'BioSample', 'Count': '0', 'Status': 'Term or Database is not found'}, {'DbName': 'biocollections', 'MenuName': 'BioCollections', 'Count': '0', 'Status': 'Term or Database is not found'}]\n",
      "115\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"cwarner@rockefeller.edu\"\n",
    "\n",
    "handle = Entrez.egquery(term=\"Opuntia AND rpl16\")\n",
    "record = Entrez.read(handle)\n",
    "print(record.keys())\n",
    "print(record[\"eGQueryResult\"])\n",
    "\n",
    "for row in record['eGQueryResult']:\n",
    "    if row[\"DbName\"] == \"nuccore\":\n",
    "        print(row[\"Count\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['Count', 'RetMax', 'RetStart', 'IdList', 'TranslationSet', 'TranslationStack', 'QueryTranslation'])\n",
      "['2689244941', '2689244853', '2666708278', '2666703263', '2419852361', '2643614054', '2627891212', '2627891124', '2627891047', '2627890961', '2627890885', '2627890802', '2627890718', '2627890636', '2627890548', '2627890460', '2627890382', '2627890305', '2627890228', '2627890152', '2627890068', '2627889984', '2627889900', '2627889817', '2627889740', '2627889653', '2627889560', '2627889470', '2627889380', '2627889278', '2627889193', '2627889118', '2627889035', '2627888952', '2582771976', '2582771888', '2582771811', '2582771725', '2582771649', '2582771566', '2582771482', '2582771400', '2582771312', '2582771241', '2582771153', '2582771075', '2582770998', '2582770921', '2582770845', '2582770761', '2582770677', '2582770593', '2582770510', '2582770433', '2582770349', '2582770274', '2582770192', '2582770117', '2582770042', '2582769959', '2582769885', '2582769808', '2582769733', '2582769650', '2582769567', '2496231309', '2281558632', '2438313908', '2308289420', '2308289413', '2308289406', '2308289399', '2308289392', '2308289385', '2308289378', '2308289371', '2272655797', '2261459663', '1972904692', '1972904685', '1972904678', '1972904671', '1972904664', '1972904657', '1972904650', '1972904643', '1972904636', '1972904629', '1972904622', '1841709044', '377581039', '330887241', '330887240', '330887239', '330887238', '330887237', '330887236', '330887235', '330887233', '330887232', '330887231', '330887228', '330887227', '330887226', '330887225', '330887224', '57240072', '57240071', '6273287', '6273291', '6273290', '6273289', '6273286', '6273285', '6273284']\n",
      "115\n"
     ]
    }
   ],
   "source": [
    "handle = Entrez.esearch(db = \"nuccore\", term = \"Opuntia AND rpl16\", retmax = 200)\n",
    "record = Entrez.read(handle)\n",
    "print(record.keys())\n",
    "id_list = record[\"IdList\"]\n",
    "print(id_list)\n",
    "print(len(id_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "handle = Entrez.efetch(db = \"nuccore\", id = id_list, rettype = \"gb\", retmore = \"text\")\n",
    "records = SeqIO.parse(handle, format = \"genbank\")\n",
    "\n",
    "with open(\"entrez-data/opuntia.gb\", \"w\") as output_handle:\n",
    "    for record in records:\n",
    "        SeqIO.write(record, output_handle, \"genbank\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GACCAAACAG\n",
      "TTTGTTGAAG\n",
      "GCGAACGACG\n",
      "ATAAATAATT\n",
      "ATAAATAATT\n",
      "GCGAACGACG\n",
      "ACTTAATAGC\n",
      "AAAAAGAAAT\n",
      "TTAGAAAGAA\n",
      "GAAAGGGTAG\n",
      "ACAGTAAGAA\n",
      "CCAAGTCAAG\n",
      "CCAAGTCAAG\n",
      "CCAAGTCAAG\n",
      "AAAAAGAAAT\n",
      "AAAAAGAAAT\n",
      "AAAAAGAAAT\n",
      "AAGAATTGAA\n",
      "TTAGAAAGAA\n",
      "TTAGAAAGAA\n",
      "CCAAGTCAAG\n",
      "CTTGCGCCAA\n",
      "CTTGCGCCAA\n",
      "CTTGCGCCAA\n",
      "AAGAATTGAA\n",
      "CTTGCGCCAA\n",
      "CCAAGTCAAG\n",
      "TTAGAAAGAA\n",
      "TTAGAAAGAA\n",
      "CTTGCGCCAA\n",
      "AAGAATTGAA\n",
      "TTAGAAAGAA\n",
      "CTTGCGCCAA\n",
      "CCAAGTCAAG\n",
      "ACTTAATAGC\n",
      "AAAAAGAAAT\n",
      "TTAGAAAGAA\n",
      "GAAAGGGTAG\n",
      "ACAGTAAGAA\n",
      "CCAAGTCAAG\n",
      "CCAAGTCAAG\n",
      "CCAAGTCAAG\n",
      "AAAAAGAAAT\n",
      "GGCGAACGAC\n",
      "AAAAAGAAAT\n",
      "AAAAAGAAAT\n",
      "AAGAATTGAA\n",
      "TTAGAAAGAA\n",
      "TTAGAAAGAA\n",
      "CCAAGTCAAG\n",
      "CTTGCGCCAA\n",
      "CTTGCGCCAA\n",
      "CTTGCGCCAA\n",
      "AAGAATTGAA\n",
      "CTTGCGCCAA\n",
      "CGAGAAAGGG\n",
      "CCAAGTCAAG\n",
      "TTAGAAAGAA\n",
      "TTAGAAAGAA\n",
      "CTTGCGCCAA\n",
      "ATTTACAGAC\n",
      "AAGAATTGAA\n",
      "TTAGAAAGAA\n",
      "CTTGCGCCAA\n",
      "CCAAGTCAAG\n",
      "CTTAGTGTGT\n",
      "ATACTTTCAA\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATACTTTCAA\n",
      "GTAAGAGCCC\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "ATGCTTAGTG\n",
      "TTCTATAAAC\n",
      "AACCCCAAAA\n",
      "GTGATATACG\n",
      "TGATATACGA\n",
      "GCCCATAGTA\n",
      "GATATACGAA\n",
      "AAAAGTAAGA\n",
      "TTGTGNGNCT\n",
      "GGGCCCNNNA\n",
      "GAGCCCATAG\n",
      "CCGNNCNTTG\n",
      "GTAAGAGCCC\n",
      "ATATACGAAA\n",
      "GATATACGAA\n",
      "GTGATATCGA\n",
      "GCTGTGATAT\n",
      "TGATATACGA\n",
      "CATTAAAGAA\n",
      "CATTAAAGGA\n",
      "TATACATTAA\n",
      "TATACATTAA\n",
      "TATACATTAA\n",
      "TATACATTAA\n",
      "TATACATAAA\n",
      "TATACATTAA\n",
      "TATACATTAA\n"
     ]
    }
   ],
   "source": [
    "records = SeqIO.parse(\"entrez-data/opuntia.gb\", format = \"genbank\")\n",
    "for record in records:\n",
    "    print(record.seq[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Useful Links and Documentation\n",
    "\n",
    "* [Bio.SearchIO Documentation](https://biopython.org/docs/latest/api/Bio.SearchIO.html)\n",
    "* [Bio.Entrez Documentation](https://biopython.org/docs/latest/api/Bio.Entrez.html)\n",
    "* [Entrez Help](https://www.ncbi.nlm.nih.gov/books/NBK3837/)\n",
    "\n",
    "## Credits and Inspiration\n",
    "\n",
    "* [Biopython Tutorial and Cookbook](https://biopython.org/DIST/docs/tutorial/Tutorial.html)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
